#!/usr/bin/env python3
"""
Servidor de Notifica√ß√µes - Implementa√ß√£o gRPC Otimizada
"""

import sys
import os
import grpc
from concurrent import futures
import logging

# Configura√ß√£o de path
sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

# Imports gRPC
from grpc_services import services_pb2, services_pb2_grpc

# Configura√ß√£o de logging
logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')

class NotificationsServiceServicer(services_pb2_grpc.NotificationsServiceServicer):
    """Servicer para o servi√ßo de notifica√ß√µes"""
    
    def __init__(self):
        self.notifications_sent = 0
        logging.info("Servi√ßo de Notifica√ß√µes inicializado")
    
    def GetNotificacoesStatus(self, request, context):
        """Retorna status do servi√ßo"""
        return services_pb2.StatusResponse(
            message=f"Notifica√ß√µes ativas - {self.notifications_sent} notifica√ß√µes enviadas"
        )
    
    def SendNotification(self, request, context):
        """Envia notifica√ß√£o para usu√°rio"""
        try:
            self.notifications_sent += 1
            
            # Simula envio de notifica√ß√£o
            notification_id = f"notif_{self.notifications_sent:06d}"
            
            logging.info(f"üì§ Notifica√ß√£o enviada para usu√°rio {request.id_usuario}: {request.message[:50]}...")
            
            return services_pb2.SendNotificationResponse(
                message=f"Notifica√ß√£o enviada com sucesso [ID: {notification_id}]"
            )
            
        except Exception as e:
            logging.error(f"‚ùå Erro ao enviar notifica√ß√£o: {e}")
            context.set_code(grpc.StatusCode.INTERNAL)
            context.set_details(f"Erro interno: {str(e)}")
            return services_pb2.SendNotificationResponse(
                message=f"Erro ao enviar notifica√ß√£o: {str(e)}"
            )

def serve():
    """Inicia o servidor de notifica√ß√µes"""
    port = 8084
    server = grpc.server(futures.ThreadPoolExecutor(max_workers=10))
    
    # Adiciona o servicer
    services_pb2_grpc.add_NotificationsServiceServicer_to_server(
        NotificationsServiceServicer(), server
    )
    
    # Configura porta
    listen_addr = f'localhost:{port}'
    server.add_insecure_port(listen_addr)
    
    try:
        server.start()
        logging.info(f"Servidor de Notifica√ß√µes rodando em {listen_addr}")
        server.wait_for_termination()
    except KeyboardInterrupt:
        logging.info("Parando servidor de notifica√ß√µes...")
        server.stop(0)
    except Exception as e:
        logging.error(f"‚ùå Erro no servidor: {e}")

if __name__ == '__main__':
    serve()